### PR TITLE
Issue #476: Add email (field) module to core.

### PR DESCRIPTION
**Pull request by hellomrcat**
PR for https://github.com/backdrop/backdrop-issues/issues/476

### CODE REVIEW

File: core/modules/email/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Email Field
+======================
+
+This module provides an email field type. This project is a Backdrop CMS version of the [Email Field](https://www.drupal.org/project/email) Drupal module, created by Matthias Hutterer.
+
+Installation
+-------------
+
+1. Place the entire email directory into your Backdrop CMS modules/ directory.
+
```
Comment (docwilmot) [inline core/modules/email/README.md]: this would need to change.

File: core/modules/email/email.feeds.inc [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * @file
+ * Integration with the Feeds module.
+ */
```
Comment (docwilmot) [inline core/modules/email/email.feeds.inc]: Dont need this either I'd think

File: core/modules/email/email.migrate.inc [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/**
+ * @file
+ * Support for the Migrate package.
+ */
```
Comment (docwilmot) [inline core/modules/email/email.migrate.inc]: Nor this

**Comment by docwilmot** [comment]
And I suspect we should also be replacing any fields in core which store emails with email field code?

File: core/modules/email/email.info [hunk 4]
```diff
@@ -0,0 +1,6 @@
+name = Email
+description = Defines an email field type.
+backdrop = 1.x
```
Comment (docwilmot) [inline core/modules/email/email.info]: this should be BACKDROP_VERSION

File: core/modules/email/email.module [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+      'description' => t('This field stores and renders email addresses.'),
+      'default_widget' => 'email_textfield',
+      'default_formatter' => 'email_default',
+      'property_type' => 'text',
+    ),
+  );
+}
+
+/**
+ * Implements hook_migrate_api().
```
Comment (docwilmot) [inline core/modules/email/email.module]: this should go

File: core/modules/email/email.feeds.inc [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Implements hook_feeds_processor_targets_alter().
```
Comment (docwilmot) [inline core/modules/email/email.feeds.inc]: Nor this

File: core/modules/email/email.module [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+ */
+function email_admin_settings_submit($form, &$form_state) {
+  $config = config('email.settings');
+  $config->set('hourly_threshold', $form_state['values']['email_hourly_threshold']);
+  $config->save();
+  backdrop_set_message(t('The configuration options have been saved.'));
+}
+
+/**
+ * Implements hook_content_migrate_instance_alter().
```
Comment (docwilmot) [inline core/modules/email/email.module]: And this

File: core/modules/email/email.info [hunk 8]
```diff
@@ -0,0 +1,6 @@
+name = Email
+description = Defines an email field type.
+BACKDROP_VERSION = 1.x
```
Comment (docwilmot) [inline core/modules/email/email.info]: Sorry, it should be core = BACKDROP_VERSION

File: core/modules/email/email.info [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+package = Fields
+
+files[] = email.migrate.inc
```
Comment (docwilmot) [inline core/modules/email/email.info]: This should be removed.

File: core/modules/email/email.module [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+function email_config_info() {
+  $prefixes['email.settings'] = array(
+    'label' => t('Email settings'),
+    'group' => t('Configuration'),
+  );
+  return $prefixes;
+}
+
+
+//TODO Token support
```
Comment (docwilmot) [inline core/modules/email/email.module]: Token is in core now, so should be supported

**Comment by docwilmot** [comment]
Sorry to be annoying here but there is still support for Feeds, Migrate and Diff in your PR. I dont think core usually includes code for supporting contrib modules.

edit: I also dont think we need the contact form code.

**Comment by docwilmot** [comment]
Incidentally, https://www.drupal.org/files/drupal8.email-field.29.patch was the Drupal patch that got merged into D8, and it also comes with its own tests, which this PR doesnt have yet; maybe it would be easier to use that (or at least match its code) for our implementation? Not sure if we would have any major differences.

**Comment by hellomrcat** [comment]
@docwilmot thanks for all the feedback. Should I just get rid the `email.feeds.inc`, `email.diff.inc`, and `email.migrate.inc` files?

**Comment by docwilmot** [comment]
> Should I just get rid the email.feeds.inc, email.diff.inc, and email.migrate.inc files

Yes, I think you should. 

Also, have a look at the D7 email patch I linked to; very simple, using HTML5's email element instead of having custom validation (the email.module uses the valid_email_address() function to validate; this shouldnt really be necessary now since the browser will do that). 

Your PR defines 4 different field formatters as well (email_default, email_plain, email_contact, email spamspan) when really we should only need one (D8 called it email_mailto). The spamspan stuff definitely should go.

In summary, I thnk just the code from the D8 patch should be ideal, unless @quicksketch disagrees.

And finally, I should have just paid more attention and suggested all this before you had to do 3 more PRs, sorry.

**Comment by quicksketch** [comment]
> Also, have a look at the D7 email patch I linked to; very simple, using HTML5's email element instead of having custom validation (the email.module uses the valid_email_address() function to validate; this shouldnt really be necessary now since the browser will do that)

Just a note that we still need server-side validation. Browser validation can easily be circumvented or on older browsers not present at all.

I'll take a look at this and see what we can cut. Great job so far, and thanks so much @docwilmot for the extensive reviews!

**Comment by hellomrcat** [comment]
@quicksketch do you want to keep all 4 field formats in the original module, or simplify to fewer?

File: core/modules/email/README.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+License
+--------
+
+This project is GPL v2 software. See the LICENSE.txt file in this directory for complete text.
+
+Maintainers
+------------
+
+Juan Olalla (https://github.com/juanolalla/)
+
```
Comment (quicksketch) [inline core/modules/email/README.md]: Let's remove this entire file. We don't include README.md files for core modules.

File: core/modules/email/email.install [hunk 12]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+
```
Comment (quicksketch) [inline core/modules/email/email.install]: Minor, let's remove the extra returns from the top of all files. A long time ago Drupal included CVS version strings at the top of files, but that hasn't been the case in a long time. The top of files should be:

```
<?php
/**
 * @file
 * Description of file.
 */
```

File: core/modules/email/email.install [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Implements hook_field_schema().
+ */
+function email_field_schema($field) {
+  return array(
+    'columns' => array(
+      'email' => array(
+        'type' => 'varchar',
+         'length' => 255,
+         'not null' => FALSE,
```
Comment (quicksketch) [inline core/modules/email/email.install]: There's one extra white space in front of 'length' and 'not null'.

**Comment by quicksketch** [comment]
Thanks again @hellomrcat! I've filed a new PR fixing the tests and trimming down functionality a bit at https://github.com/backdrop/backdrop/pull/1029.

More info in the main thread: https://github.com/backdrop/backdrop-issues/issues/476
